File: multiGMRES.m

function main
  
  printf("Loading AA matrix\n");
 
  load AA.dat
%   fd = fopen('AA.dat','r');
%   [i,j,val] = fscanf(fd,'%d %d %f');
%   fclose(fd);
  

  printf("Converting AA matrix\n");
  AA = spconvert(AA);
  %AA = sparse(i,j,val)
  printf("AA Matrix ready\n");
  
  %display (AA)
  
  fd = fopen('bb.dat','r')
  bb = fscanf(fd,'%f');
  fclose(fd);


  xx = zeros(size(bb,1),1);
  MM = speye(size(bb,1));


  display(size(AA))
  display(size(xx))
  display(size(bb))

  printf("Building Krylov space\n");
  
  MM = inv(diag(diag(AA)));
  
  %display (MM)
  
  krylovSpaceNorms( AA, xx, bb, MM, 10);
  
  %display (xx)
  
  %display (A*xx - b)
  
  %display norm(A*xx - b)

end

function [x] = krylovSpaceNorms ( A, x, b, M, n )

  bNorm = norm(b);
  printf("Initial Residual: %e\n",norm(A*x - b)/bNorm)
  
  
  for it = 1:100
    
    res = b - A*x;
    
    k{1} = res;
    
    %printf ( '%d, %e\n',1,norm(k{1}));
    for i = 2:n+1,
        t = M*k{i-1};
        k{i} = A*t;
        %printf( '%d, %e\n',i,norm(k{i}));
    end
    
    %display("Building dot product matrix\n");
    
    for i = 2:n+1,
        r(i-1) = dot(k{i},res);
        for j = 2:n+1,
            N(i-1,j-1) = dot(k{i},k{j});
        end
    end
    
    %display(det(N))
    %display(N)
    %display(r')
    
    alpha = N \ r';
    
    %display (alpha)
    
    err = N*alpha - r';
    %display (norm(err))

%    dalpha = N \ (err);
%    alpha -= dalpha;
%    err = N*alpha - r';
%    display (norm(err))
    
%    dalpha = N \ (err);
%    alpha -= dalpha;
%    err = N*alpha - r';
%    display (norm(err))   


    y = zeros(size(x));
    for i = 1:n
      y += alpha(i)*k{i};
    end
    x += M*y;
    
    display (norm(A*x-b)/bNorm)
    %display(x)
    
  end

end

main

File: multiGMRES.py
""" this is a multi-basis vector version of GMRES"""

import csv
import scipy.sparse
from numpy import diag, zeros, zeros_like, loadtxt, array, matrix, dot
from numpy.linalg import inv, norm, det, solve

def multiGMRES():
    "multi basis vector GMRES variant"

    print("Loading matrices...")

    rows = []
    cols = []
    vals = []

    # AA = scipy.sparse.dok_matrix((100000, 100000))
    csvreader = csv.reader(open('AA.dat'), delimiter=' ', skipinitialspace=True)
    for line in csvreader:
        row, column, val = line[:3]
        rows.append(int( row ))
        cols.append(int( column ))
        vals.append(float( val ))
        # AA[int(row), int(column)] = float(val)

    rows = array(rows)-1
    cols = array(cols)-1
    vals = array(vals)
    AA = scipy.sparse.coo_matrix((vals,( rows, cols )))
    AA = AA.tocsr()

    bb = loadtxt(open('bb.dat'))
    nn = len(bb)

    bb = matrix(bb).transpose()

    print("Matrices loaded.")

    xx = zeros((len(bb),1))
    MM = scipy.sparse.identity(len(bb))

    print("Building Krylov space")
    inv_diag = 1./AA.diagonal()
    MM = scipy.sparse.spdiags(inv_diag, [0,], nn, nn)

    krylovSpaceNorms(AA, xx, bb, MM, 10)

def krylovSpaceNorms(A, x, b, M, n):
    "do multiGMRES"

    k = []
    nn = len(b)
    for _ in range(n+1):
        kvec = array((nn, 1))
        kvec.fill(333333.)
        k.append(kvec)

    bNorm = norm(b, ord=2)

    print("Initial Residual: ", norm(A*x - b)/bNorm)

    for _ in range(0, 100):

        print("ITERATION ", _)

        res = b - A*x

        # construct n+1 krylov vectors, not using the first to make N
        k[0] = res

        for i in range(1, n+1):
            print("i = ", i)
            t = M*k[i-1]
            k[i] = A*t

        print("Building dot product matrix\n")

        r = zeros(n)
        r.fill(4444444.)
        N = zeros((n, n))
        N.fill(8888888.)
        for i in range(1, n+1):
            r[i-1] = dot(k[i].transpose(), res)
            for j in range(1, n+1):
                N[i-1, j-1] = dot(k[i].transpose(), k[j-1])

        alpha = solve(N, r)

        # err = N*alpha - r
        # dalpha = solve(N, err)
        # alpha -= dalpha
        # err = N*alpha - r.transpose()

        y = zeros_like(x)
        for i in range(1, n):
            y += alpha[i]*k[i]
        x += M*y

        print(norm(A*x-b)/bNorm)

multiGMRES()
